### PR TITLE
Fix restored optimizer learning rates in TorchFT

### PR DESCRIPTION
**Pull request by jojoinfra**
## Summary

While performing zero-deviation numerical alignment validation with TorchTitan + TorchFT, I found that a joining replica could resume with an incorrect **learning rate**. Although gradient all-reduce produced identical gradients across replicas, different learning rates caused their weights to diverge after the optimizer update.

This breaks a fundamental invariant of synchronous data-parallel training: replicas starting from identical parameters and optimizer state must apply identical updates after gradient synchronization.

## Root cause

Two issues contribute to the incorrect learning rate:

- The optimizer state dictionary is cached. Tensor state remains current through shared references, but scalar learning rates become stale when the scheduler updates the optimizer.
- Restoring scheduler state updates scheduler metadata without updating the actual optimizer parameter groups. Consequently, the scheduler can report the expected learning rate while the optimizer uses a different value.

torchft_lr_bug_root_cause

An example:

torchft_lr_example

## Changes

- Refresh learning rates in the cached optimizer state dictionary before exporting it.
- Restore actual optimizer learning rates from the scheduler after both optimizer and scheduler states have finished loading, covering online replica recovery and DCP checkpoint loading without advancing the scheduler.
- Add one focused regression test asserting that a joining replica restores the same actual optimizer learning rate as a healthy replica.

### CODE REVIEW

File: torchtitan/experiments/torchft/checkpoint.py [hunk 1]
```diff
@@ -115,6 +116,8 @@ def __init__(
             optimizers.init_cache_state_dict()
 
             def state_dict():
+                if isinstance(optimizers, TorchFTOptimizersContainer):
```
Comment (tianyu-l) [inline torchtitan/experiments/torchft/checkpoint.py]: do you need this `if` -- this checkpoint component is not supposed to be used without the ft optimizer, so maybe `assert`

Comment (jojoinfra) [reply torchtitan/experiments/torchft/checkpoint.py]: updated to use an assert

**Review by tianyu-l — APPROVED** [review]
